README.md
# TemplateEngineTwig

[![Build Status](https://travis-ci.org/wanze/TemplateEngineTwig.svg?branch=master)](https://travis-ci.org/wanze/TemplateEngineTwig)
[![StyleCI](https://github.styleci.io/repos/21304492/shield?branch=master)](https://github.styleci.io/repos/21304492)
[![License: MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/licenses/MIT)
[![ProcessWire 3](https://img.shields.io/badge/ProcessWire-3.x-orange.svg)](https://github.com/processwire/processwire)

A ProcessWire module adding Twig to the [TemplateEngineFactory](https://github.com/wanze/TemplateEngineFactory).

## Requirements

* ProcessWire `3.0` or newer
* TemplateEngineFactory `2.0` or newer
* PHP >= `7.0` for version `2.x`, PHP >= `7.2.5` for version `3.x`
* Composer

> The `1.x` version of this module is available on the [1.x branch](https://github.com/wanze/TemplateEngineTwig/tree/1.x).
Use this version if you still use _TemplateEngineFactory_ `1.x`.  

## Installation

Execute the following command in the root directory of your ProcessWire installation:

```
composer require wanze/template-engine-twig:^3.0
```

This will install the _TemplateEngineTwig_ and _TemplateEngineFactory_ modules in one step. Afterwards, don't forget 
to enable Twig as engine in the _TemplateEngineFactory_ module's configuration.

> ℹ️ This module includes test dependencies. If you are installing on production with `composer install`, make sure to
pass the `--no-dev` flag to omit autoloading any unnecessary test dependencies!.

## Configuration

The module offers the following configuration:

* **`Template files suffix`** The suffix of the Twig template files, defaults to `twig.html`.
* **`Provide ProcessWire API variables in Twig templates`** API variables (`$pages`, `$input`, `$config`...)
are accessible in Twig,
e.g. `{{ config }}` for the config API variable.
* **`Debug`** If enabled, Twig outputs debug information. The module also registers the _Debug Extension_, offering
the `{{ dump() }}` function to inspect variables.  
* **`Auto reload templates (recompile)`** If enabled, templates are recompiled whenever the source code changes.
* **`Strict variables`** If set to `false`, Twig will silently ignore invalid variables (variables and
or attributes/methods that do not exist) and replace them with a `null` value. When set to `true`,
Twig throws an exception instead
* **`Auto escape variables`** If enabled, templates will auto-escape variables. If you are using ProcessWire
textformatters to escape field values, do not enable this feature.

## Extending Twig

It is possible to extend Twig after it has been initialized by the module. Hook the method `TemplateEngineTwig::initTwig`
to register custom functions, extensions, global variables, filters etc.

Here is an example how you can use the provided hook to attach a custom function.

```php
wire()->addHookAfter('TemplateEngineTwig::initTwig', function (HookEvent $event) {
    /** @var \Twig\Environment $twig */
    $twig = $event->arguments('twig');

    $twig->addFunction(new \Twig\TwigFunction('processwire', function () {
        return 'ProcessWire rocks!';
    }));
});

// ... and then use it anywhere in a Twig template:

{{ processwire() }}
```

> The above hook can be put in your `site/init.php` file. If you prefer to use modules, put it into the module's `init()`
method and make sure that the module is auto loaded.

### Use Twig Extensions

> The [Twig Extensions](https://twig-extensions.readthedocs.io/en/latest/) is a library that provides several useful
extensions for Twig.

These extensions are not included in the module by default, but you can add them to Twig using the same hook explained
above.

Fist, install the library with Composer in the ProcessWire root directory: `composer require twig/extensions`. Next,
use the same hook as above to register the desired extensions.

```php
wire()->addHookAfter('TemplateEngineTwig::initTwig', function (HookEvent $event) {
    /** @var \Twig\Environment $twig */
    $twig = $event->arguments('twig');
    
    // Register the extensions.
    $twig->addExtension(new Twig_Extensions_Extension_Text());
    $twig->addExtension(new Twig_Extensions_Extension_I18n());
    $twig->addExtension(new Twig_Extensions_Extension_Intl());
});
```


CHANGELOG.md
# Changelog

## [Unreleased]


## [3.0.1] - 2023-08-03

### Fixed

* Deprecation warnings for PHP 8.2 - thanks @hiboudev (see [#29](https://github.com/wanze/TemplateEngineTwig/pull/23))
* Add support for Composer 2

## [3.0.0] - 2019-02-01

* Update Twig to version `3.x` - thanks @porl (see [#23](https://github.com/wanze/TemplateEngineTwig/pull/23))

Twig 3 uses namespaced classes and requires PHP `^7.2.5`. Make sure to test your installation after updating,
especially if you use hooks to customize Twig.

## [2.0.0] - 2018-02-13

* Initial release of the new major `2.x` version 🐣

### Added

* Add a setting to enable debug mode, independently from ProcessWire's debug mode.  
* The `\Twig_Extension_Debug` is automatically added if debug mode is active.
* Add unit tests 🎉

### Changed

* The Twig dependencies are no longer part of this repository. Installation is therefore only
possible with Composer, no longer via ProcessWire modules directory. 
* Update Twig to version `2.6.2`. Because of that, PHP > `7.x` is required.

## [1.1.0] - 2018-03-19

### Changed

* Update twig to latest version `1.35.2`.
* Define this repository as ProcessWire module in `composer.json` so it can be installed via composer
([more information](http://harikt.com/blog/2013/11/16/composer-support-for-processwire-modules/)).

### Fixed

* Fix autoloader deprecation message - by @lesaff.
* Make sure that some configuration settings are correctly passed as booleans to twig - by @nextgensparx

[3.0.1]: https://github.com/wanze/TemplateEngineTwig/releases/tag/v3.0.1
[3.0.0]: https://github.com/wanze/TemplateEngineTwig/releases/tag/v3.0.0
[2.0.0]: https://github.com/wanze/TemplateEngineTwig/releases/tag/v2.0.0
[1.1.0]: https://github.com/wanze/TemplateEngineTwig/releases/tag/v1.1.0
[Unreleased]: https://github.com/wanze/TemplateEngineTwig/compare/v3.0.0...HEAD


tests/bootstrap.php
<?php

require_once(dirname(dirname(dirname(dirname(__DIR__)))) . '/index.php');

// Install TemplateEngineFactory & TemplateEngineTwig.
// Note: We use get() instead of install() which will install the modules and call the init() method!
$wire->wire('modules')->get('TemplateEngineFactory');
$wire->wire('modules')->get('TemplateEngineTwig');


tests/TemplateEngineTwigTest.php
<?php

namespace TemplateEngineTwig\Test;

use PHPUnit\Framework\TestCase;
use ProcessWire\HookEvent;
use ProcessWire\ProcessWire;
use TemplateEngineTwig\TemplateEngineTwig;
use Twig_Error_Loader;

class TemplateEngineTwigTest extends TestCase
{
    /**
     * @var \ProcessWire\ProcessWire
     */
    private $wire;

    protected function setUp()
    {
        $this->bootstrapProcessWire();
        $this->fakeSitePath();
    }

    protected function tearDown()
    {
        ProcessWire::removeInstance($this->wire);
    }

    /**
     * @dataProvider initializeTwigDataProvider
     */
    public function test_twig_initialized_correctly(array $moduleConfig, $autoReload, $strictVars, $debug, $wireDebug)
    {
        $this->wire->wire('config')->debug = $wireDebug;

        $this->wire->addHookAfter('TemplateEngineTwig::initTwig',
            function (HookEvent $event) use ($autoReload, $strictVars, $debug) {
                /** @var \Twig_Environment $twig */
                $twig = $event->arguments('twig');

                $this->assertEquals($debug, $twig->isDebug());
                $this->assertEquals($autoReload, $twig->isAutoReload());
                $this->assertEquals($strictVars, $twig->isStrictVariables());
            });

        $engine = $this->getTwigEngine($moduleConfig);

        $engine->render('dummy');
    }

    public function test_hook_initTwig()
    {
        $this->wire->addHookAfter('TemplateEngineTwig::initTwig',
            function (HookEvent $event) {
                /** @var \Twig_Environment $twig */
                $twig = $event->arguments('twig');

                $twig->addFunction(new \Twig_Function('processwire', function () {
                    return 'ProcessWire rocks!';
                }));
            });

        $engine = $this->getTwigEngine();

        $this->assertEquals("ProcessWire rocks!\n", $engine->render('test_custom_function'));
    }

    public function test_missing_template_throws_exception()
    {
        $engine = $this->getTwigEngine();

        $this->expectException(Twig_Error_Loader::class);

        $engine->render('this/template/does/not/exist');
    }

    public function test_rendering_with_or_without_templates_suffix()
    {
        $engine = $this->getTwigEngine();

        $this->assertEquals($engine->render('dummy'), $engine->render('dummy.html.twig'));
    }

    public function test_rendering_api_variables_available()
    {
        $engine = $this->getTwigEngine();

        $this->assertEquals("API variables available.", $engine->render('test_api_vars'));
    }

    public function test_rendering_template_in_subfolder()
    {
        $engine = $this->getTwigEngine();

        $this->assertEquals("Dummy in subfolder!\n", $engine->render('subfolder/dummy'));
    }

    public function test_rendering_custom_template_files_suffix()
    {
        $engine = $this->getTwigEngine(['template_files_suffix' => 'processwire.twig']);

        $this->assertEquals("ProcessWire!\n", $engine->render('test_custom_suffix'));
        $this->assertEquals("ProcessWire!\n", $engine->render('test_custom_suffix.processwire.twig'));
    }

    public function test_rendering_data()
    {
        $engine = $this->getTwigEngine();

        $data = [
            'Breaking Bad',
            'Sons of Anarchy',
            'Big Bang Theory',
        ];

        $this->assertEquals(implode(',', $data) . "\n", $engine->render('test_data', ['data' => $data]));
    }

    /**
     * @return array
     */
    public function initializeTwigDataProvider()
    {
        return [
            [
                [
                    'template_files_suffix' => 'html.twig',
                    'api_vars_available' => 1,
                    'auto_reload' => 1,
                    'auto_escape' => 0,
                    'strict_variables' => 0,
                    'debug' => 'config',
                ],
                true,   // auto reload
                false,  // strict variables
                false,  // debug
                false,  // $config->debug
            ],
            [
                [
                    'template_files_suffix' => 'html.twig',
                    'api_vars_available' => 1,
                    'auto_reload' => 1,
                    'auto_escape' => 0,
                    'strict_variables' => 0,
                    'debug' => 'config',
                ],
                true,   // auto reload
                false,  // strict variables
                true,   // debug
                true,   // $config->debug
            ],
            [
                [
                    'template_files_suffix' => 'html.twig',
                    'api_vars_available' => 1,
                    'auto_reload' => 0,
                    'auto_escape' => 0,
                    'strict_variables' => 1,
                    'debug' => 0,
                ],
                false,
                true,
                false,
                true,
            ],
            [
                [
                    'template_files_suffix' => 'html.twig',
                    'api_vars_available' => 1,
                    'auto_reload' => 0,
                    'auto_escape' => 0,
                    'strict_variables' => 1,
                    'debug' => 1,
                ],
                false,
                true,
                true,
                false,
            ]
        ];
    }

    /**
     * Let $config->paths->site point to this directory.
     */
    private function fakeSitePath()
    {
        $paths = $this->wire->wire('config')->paths;
        $paths->set('site', 'site/modules/TemplateEngineTwig/tests/');
    }

    /**
     * @param array $factoryConfig
     * @param array $moduleConfig
     *
     * @return \TemplateEngineTwig\TemplateEngineTwig
     */
    private function getTwigEngine(array $moduleConfig = [], array $factoryConfig = [])
    {
        $factoryConfig = array_merge($this->getFactoryConfig(), $factoryConfig);
        $moduleConfig = array_merge($this->getModuleConfig(), $moduleConfig);

        return new TemplateEngineTwig($factoryConfig, $moduleConfig);
    }

    private function getModuleConfig()
    {
        return [
            'template_files_suffix' => 'html.twig',
            'api_vars_available' => 1,
            'auto_reload' => 1,
            'auto_escape' => 0,
            'strict_variables' => 0,
            'debug' => 'config',
        ];
    }

    private function getFactoryConfig()
    {
        return $this->wire->wire('modules')->get('TemplateEngineFactory')->getArray();
    }

    private function bootstrapProcessWire()
    {
        $rootPath = __DIR__ . '../../../../../';
        $config = ProcessWire::buildConfig($rootPath);
        $this->wire = new ProcessWire($config);
    }
}


tests/src/TemplateEngineTwigTest.php
<?php

namespace TemplateEngineTwig\Test;

use ProcessWire\HookEvent;
use TemplateEngineTwig\TemplateEngineTwig;
use Twig\Environment;
use Twig\Error\LoaderError;
use Twig\TwigFunction;

/**
 * Tests for the TemplateEngineTwig class.
 *
 * @coversDefaultClass \TemplateEngineTwig\TemplateEngineTwig
 *
 * @group TemplateEngineTwig
 */
class TemplateEngineTwigTest extends ProcessWireTestCaseBase
{
    protected function setUp()
    {
        parent::setUp();

        $this->fakePath('site', 'site/modules/TemplateEngineTwig/tests/');
    }

    /**
     * @test
     * @dataProvider initializeTwigDataProvider
     */
    public function it_should_initialize_twig_correctly_depending_on_module_configuration(array $moduleConfig, $autoReload, $strictVars, $debug, $wireDebug)
    {
        $this->wire('config')->debug = $wireDebug;

        $this->addHookAfter('TemplateEngineTwig::initTwig',
            function (HookEvent $event) use ($autoReload, $strictVars, $debug) {
                /** @var Environment $twig */
                $twig = $event->arguments('twig');

                $this->assertEquals($debug, $twig->isDebug());
                $this->assertEquals($autoReload, $twig->isAutoReload());
                $this->assertEquals($strictVars, $twig->isStrictVariables());
            });

        $engine = $this->getTwigEngine($moduleConfig);

        // Twig is only initialized when we actually render something.
        $engine->render('dummy');
    }

    /**
     * @test
     */
    public function it_should_be_possible_to_extend_twig()
    {
        $this->addHookAfter('TemplateEngineTwig::initTwig',
            function (HookEvent $event) {
                /** @var Environment $twig */
                $twig = $event->arguments('twig');

                $twig->addFunction(new TwigFunction('processwire', function () {
                    return 'ProcessWire rocks!';
                }));
            });

        $engine = $this->getTwigEngine();

        $this->assertEquals("ProcessWire rocks!\n", $engine->render('test_custom_function'));
    }

    /**
     * @test
     * @covers ::render
     */
    public function it_should_throw_an_exception_if_a_template_does_not_exist()
    {
        $engine = $this->getTwigEngine();

        $this->expectException(LoaderError::class);

        $engine->render('this/template/does/not/exist');
    }

    /**
     * @test
     * @covers ::render
     */
    public function it_should_find_templates_with_or_without_suffix_and_render_the_same_output()
    {
        $engine = $this->getTwigEngine();

        $this->assertEquals("Dummy\n", $engine->render('dummy'));
        $this->assertEquals("Dummy\n", $engine->render('dummy.html.twig'));
    }

    /**
     * @test
     * @covers ::render
     */
    public function it_should_only_provide_the_api_variables_if_enabled_in_config()
    {
        $engine = $this->getTwigEngine();
        $engine2 = $this->getTwigEngine(['api_vars_available' => false]);

        $this->assertEquals("API variables available.", $engine->render('test_api_vars'));
        $this->assertNotEquals("API variables available.", $engine2->render('test_api_vars'));
    }

    /**
     * @test
     * @covers ::render
     */
    public function it_should_find_and_render_a_template_in_a_subfolder()
    {
        $engine = $this->getTwigEngine();

        $this->assertEquals("Dummy in subfolder!\n", $engine->render('subfolder/dummy'));
    }

    /**
     * @test
     * @covers ::render
     */
    public function it_should_find_and_render_a_template_with_a_custom_suffix()
    {
        $engine = $this->getTwigEngine(['template_files_suffix' => 'processwire.twig']);

        $this->assertEquals("ProcessWire!\n", $engine->render('test_custom_suffix'));
        $this->assertEquals("ProcessWire!\n", $engine->render('test_custom_suffix.processwire.twig'));
    }

    /**
     * @test
     * @covers ::render
     */
    public function it_should_pass_data_to_the_template()
    {
        $engine = $this->getTwigEngine();

        $series = [
            'Breaking Bad',
            'Sons of Anarchy',
            'Big Bang Theory',
        ];

        // The series are rendered comma separated -> {{ series|join(',') }}.
        $expected = sprintf("%s\n", implode(',', $series));

        $this->assertEquals($expected, $engine->render('test_data', ['series' => $series]));
    }

    /**
     * @return array
     */
    public function initializeTwigDataProvider()
    {
        return [
            [
                [
                    'template_files_suffix' => 'html.twig',
                    'api_vars_available' => 1,
                    'auto_reload' => 1,
                    'auto_escape' => 0,
                    'strict_variables' => 0,
                    'debug' => 'config',
                ],
                true,   // auto reload
                false,  // strict variables
                false,  // debug
                false,  // $config->debug
            ],
            [
                [
                    'template_files_suffix' => 'html.twig',
                    'api_vars_available' => 1,
                    'auto_reload' => 1,
                    'auto_escape' => 0,
                    'strict_variables' => 0,
                    'debug' => 'config',
                ],
                true,   // auto reload
                false,  // strict variables
                true,   // debug
                true,   // $config->debug
            ],
            [
                [
                    'template_files_suffix' => 'html.twig',
                    'api_vars_available' => 1,
                    'auto_reload' => 0,
                    'auto_escape' => 0,
                    'strict_variables' => 1,
                    'debug' => 0,
                ],
                false,
                true,
                false,
                true,
            ],
            [
                [
                    'template_files_suffix' => 'html.twig',
                    'api_vars_available' => 1,
                    'auto_reload' => 0,
                    'auto_escape' => 0,
                    'strict_variables' => 1,
                    'debug' => 1,
                ],
                false,
                true,
                true,
                false,
            ]
        ];
    }

    /**
     * @param array $factoryConfig
     * @param array $moduleConfig
     *
     * @return \TemplateEngineTwig\TemplateEngineTwig
     */
    private function getTwigEngine(array $moduleConfig = [], array $factoryConfig = [])
    {
        $factoryConfig = array_merge($this->getFactoryConfig(), $factoryConfig);
        $moduleConfig = array_merge($this->getModuleConfig(), $moduleConfig);

        return $this->wire(new TemplateEngineTwig($factoryConfig, $moduleConfig));
    }

    private function getModuleConfig()
    {
        return [
            'template_files_suffix' => 'html.twig',
            'api_vars_available' => 1,
            'auto_reload' => 1,
            'auto_escape' => 0,
            'strict_variables' => 0,
            'debug' => 'config',
        ];
    }

    private function getFactoryConfig()
    {
        return $this->wire('modules')->get('TemplateEngineFactory')
            ->getArray();
    }
}


src/TemplateEngineTwig.php
<?php

namespace TemplateEngineTwig;

use TemplateEngineFactory\TemplateEngineBase;
use Twig\Environment;
use Twig\Extension\DebugExtension;
use Twig\Loader\FilesystemLoader;

/**
 * Provides the Twig template engine.
 */
class TemplateEngineTwig extends TemplateEngineBase
{
    const COMPILE_DIR = 'TemplateEngineTwig_compile/';

    /**
     * @var Environment
     */
    protected $twig;

    /**
     * {@inheritdoc}
     */
    public function render($template, $data = [])
    {
        $template = $this->normalizeTemplate($template);
        $data = $this->getData($data);

        return $this->getTwig()->render($template, $data);
    }

    /**
     * @throws \ProcessWire\WireException
     *
     * @return Environment
     */
    protected function getTwig()
    {
        if ($this->twig === null) {
            return $this->buildTwig();
        }

        return $this->twig;
    }

    /**
     * @throws \ProcessWire\WireException
     *
     * @return Environment
     */
    protected function buildTwig()
    {
        $loader = new FilesystemLoader($this->getTemplatesRootPath());

        $this->twig = new Environment($loader, [
            'cache' => $this->wire('config')->paths->assets . 'cache/' . self::COMPILE_DIR,
            'debug' => $this->isDebug(),
            'auto_reload' => (bool) $this->moduleConfig['auto_reload'],
            'autoescape' => $this->moduleConfig['auto_escape'] ? 'name' : false,
            'strict_variables' => (bool) $this->moduleConfig['strict_variables'],
        ]);

        // Add the debug extension offering the "dump()" function for variables.
        if ($this->isDebug()) {
            $this->twig->addExtension(new DebugExtension());
        }

        $this->initTwig($this->twig);

        return $this->twig;
    }

    /**
     * Hookable method called after Twig has been initialized.
     *
     * Use this method to customize the passed $twig instance,
     * e.g. adding functions and filters.
     *
     * @param Environment $twig
     */
    protected function ___initTwig(Environment $twig)
    {
    }

    private function isDebug()
    {
        if ($this->moduleConfig['debug'] === 'config') {
            return $this->wire('config')->debug;
        }

        return (bool) $this->moduleConfig['debug'];
    }

    /**
     * @param array $data
     *
     * @throws \ProcessWire\WireException
     *
     * @return array
     */
    private function getData(array $data)
    {
        if (!$this->moduleConfig['api_vars_available']) {
            return $data;
        }

        foreach ($this->wire('all') as $name => $object) {
            $data[$name] = $object;
        }

        return $data;
    }

    /**
     * Normalize the given template by adding the template files suffix.
     *
     * @param string $template
     *
     * @return string
     */
    private function normalizeTemplate($template)
    {
        $suffix = $this->moduleConfig['template_files_suffix'];

        $normalizedTemplate = ltrim($template, DIRECTORY_SEPARATOR);

        if (!preg_match("/\.{$suffix}$/", $template)) {
            return $normalizedTemplate . sprintf('.%s', $suffix);
        }

        return $normalizedTemplate;
    }
}


TemplateEngineTwig.module.php
<?php

namespace ProcessWire;

use TemplateEngineTwig\TemplateEngineTwig as TwigEngine;

/**
 * Adds Twig templates to the TemplateEngineFactory module.
 */
class TemplateEngineTwig extends WireData implements Module, ConfigurableModule
{
    /**
     * @var array
     */
    private static $defaultConfig = [
        'template_files_suffix' => 'html.twig',
        'api_vars_available' => 1,
        'auto_reload' => 1,
        'auto_escape' => 0,
        'strict_variables' => 0,
        'debug' => 'config',
    ];

    public function __construct()
    {
        parent::__construct();

        $this->wire('classLoader')->addNamespace('TemplateEngineTwig', __DIR__ . '/src');
        $this->setDefaultConfig();
    }

    /**
     * @return array
     */
    public static function getModuleInfo()
    {
        return [
            'title' => 'Template Engine Twig',
            'summary' => 'Twig templates for the TemplateEngineFactory',
            'version' => 301,
            'author' => 'Stefan Wanzenried',
            'href' => 'https://processwire.com/talk/topic/6835-module-twig-for-the-templateenginefactory/',
            'singular' => true,
            'autoload' => true,
            'requires' => [
                'TemplateEngineFactory>=2.0.0',
                'PHP>=7.0',
                'ProcessWire>=3.0',
            ],
        ];
    }

    public function init()
    {
        /** @var \ProcessWire\TemplateEngineFactory $factory */
        $factory = $this->wire('modules')->get('TemplateEngineFactory');

        $factory->registerEngine('Twig', new TwigEngine($factory->getArray(), $this->getArray()));
    }

    private function setDefaultConfig()
    {
        foreach (self::$defaultConfig as $key => $value) {
            $this->set($key, $value);
        }
    }

    /**
     * @param array $data
     *
     * @throws \ProcessWire\WireException
     * @throws \ProcessWire\WirePermissionException
     *
     * @return \ProcessWire\InputfieldWrapper
     */
    public static function getModuleConfigInputfields(array $data)
    {
        /** @var Modules $modules */
        $data = array_merge(self::$defaultConfig, $data);
        $wrapper = new InputfieldWrapper();
        $modules = wire('modules');

        /** @var \ProcessWire\InputfieldText $field */
        $field = $modules->get('InputfieldText');
        $field->label = __('Template files suffix');
        $field->name = 'template_files_suffix';
        $field->value = $data['template_files_suffix'];
        $field->required = 1;
        $wrapper->append($field);

        $field = $modules->get('InputfieldCheckbox');
        $field->label = __('Provide ProcessWire API variables in Twig templates');
        $field->description = __('API variables (`$pages`, `$input`, `$config`...) are accessible in Twig, e.g. `{{ config }}` for the config API variable.');
        $field->name = 'api_vars_available';
        $field->checked = (bool) $data['api_vars_available'];
        $wrapper->append($field);

        /** @var \ProcessWire\InputfieldSelect $field */
        $field = $modules->get('InputfieldSelect');
        $field->label = __('Debug');
        $field->name = 'debug';
        $field->addOptions([
            'config' => __('Inherit from ProcessWire'),
            0 => __('No'),
            1 => __('Yes'),
        ]);
        $field->value = $data['debug'];
        $wrapper->append($field);

        /** @var \ProcessWire\InputfieldCheckbox $field */
        $field = $modules->get('InputfieldCheckbox');
        $field->label = __('Auto reload templates (recompile)');
        $field->description = __('If enabled, templates are recompiled whenever the source code changes');
        $field->name = 'auto_reload';
        $field->checked = (bool) $data['auto_reload'];
        $wrapper->append($field);

        $field = $modules->get('InputfieldCheckbox');
        $field->label = __('Strict variables');
        $field->description = __('If set to `false`, Twig will silently ignore invalid variables (variables and or attributes/methods that do not exist) and replace them with a `null` value. When set to `true`, Twig throws an exception instead');
        $field->name = 'strict_variables';
        $field->checked = (bool) $data['strict_variables'];
        $wrapper->append($field);

        $field = $modules->get('InputfieldCheckbox');
        $field->label = __('Auto escape variables');
        $field->description = __('If enabled, templates will auto-escape variables. If you are using ProcessWire textformatters to escape field values, do not enable this feature.');
        $field->name = 'auto_escape';
        $field->checked = (bool) $data['auto_escape'];
        $wrapper->append($field);

        return $wrapper;
    }
}
